File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l = 0
        r = 0
        res = 0
        myDict = {}
        while(r < len(s)):
            if s[r] not in myDict or myDict[s[r]] < l:
                myDict[s[r]] = r
            else:
                l = myDict[s[r]] + 1
                myDict[s[r]] = r
            res = max(res,r - l + 1)
            r+=1
        return res

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> map = new HashMap<>();
        int left = 0;
        int right = 0;
        int maxR = 0;
        while(right < s.length()){
            if(map.containsKey(s.charAt(right)) && map.get(s.charAt(right)) >= left){
                left = map.get(s.charAt(right)) + 1;
            }
            map.put(s.charAt(right),right);
            maxR = Math.max(maxR, right - left + 1);
            right+=1;
        }
        
        return maxR;
    }
}